toolkit/offchain/src/d_param/mod.rs

```diff
@@ -0,0 +1,316 @@
+	validator_utxos: Vec<OgmiosUtxo>,
+) -> Result<Option<(OgmiosUtxo, DParameter)>, anyhow::Error> {
+	if let Some(utxo) = validator_utxos.first() {
+		let datum = utxo.datum.clone().ok_or_else(|| {
+			anyhow!("Invalid state: an UTXO at the validator script address does not have a datum")
+		})?;
+		let datum_plutus_data = PlutusData::from_bytes(datum.bytes).map_err(|e| {
+			anyhow!("Internal error: could not decode datum of D-parameter validator script: {}", e)
+		})?;
+		let current_d_param: DParameter =
+			DParamDatum::try_from(datum_plutus_data)
+				.map_err(|e| {
+					anyhow!("Internal error: could not decode datum of D-parameter validator script: {}", e)
+				})?
+				.into();
+		Ok(Some((utxo.clone(), current_d_param)))
+	} else {
+		Ok(None)
+	}
+}
+
+async fn insert_d_param<C: QueryLedgerState + Transactions + QueryNetwork>(
+	validator: &PlutusScript,
+	policy: &PlutusScript,
+	d_parameter: &DParameter,
+	ctx: TransactionContext,
+	genesis_utxo: UtxoId,
+	client: &C,
+) -> anyhow::Result<McTxHash>
+where
+	C: Transactions,
+{
+	let zero_ex_units = ScriptExUnits {
+		mint_ex_units: vec![
+			ExUnits::new(&0u64.into(), &0u64.into()),
+			ExUnits::new(&0u64.into(), &0u64.into()),
+		],
+		spend_ex_units: vec![],
+	};
+
+	let gov_utxo = crate::init_governance::get_governance_utxo(genesis_utxo, client)
+		.await
+		.map_err(|e| JsError::from_str(e.to_string().as_str()))?;
+
+	let tx = mint_d_param_token_tx(
+		validator,
+		policy,
+		d_parameter,
+		&ctx,
+		zero_ex_units,
+		gov_utxo.clone(),
+	)?;
+	let evaluate_response = client.evaluate_transaction(&tx.to_bytes()).await.map_err(|e| {
+		anyhow!(
+			"Evaluate insert D-parameter transaction request failed: {}, bytes: {}",
+			e,
+			hex::encode(tx.to_bytes())
+		)
+	})?;
+	let mint_witness_ex_units = get_validator_budgets(evaluate_response)?;
+	let tx = mint_d_param_token_tx(
+		validator,
+		policy,
+		d_parameter,
+		&ctx,
+		mint_witness_ex_units,
+		gov_utxo,
+	)?;
+	let signed_tx = ctx.sign(&tx).to_bytes();
+	let res = client.submit_transaction(&signed_tx).await.map_err(|e| {
+		anyhow!(
+			"Submit insert D-parameter transaction request failed: {}, bytes: {}",
+			e,
+			hex::encode(signed_tx)
+		)
+	})?;
+	let tx_id = McTxHash(res.transaction.id);
+	log::info!("Transaction submitted: {}", hex::encode(tx_id.0));
+	Ok(tx_id)
+}
+
+async fn update_d_param<C: QueryLedgerState + Transactions + QueryNetwork>(
+	validator: &PlutusScript,
+	policy: &PlutusScript,
+	d_parameter: &DParameter,
+	current_utxo: &OgmiosUtxo,
+	ctx: TransactionContext,
+	genesis_utxo: UtxoId,
+	client: &C,
+) -> anyhow::Result<McTxHash>
+where
+	C: Transactions,
+{
+	let zero_ex_units = ScriptExUnits {
+		mint_ex_units: vec![ExUnits::new(&0u64.into(), &0u64.into())],
+		spend_ex_units: vec![ExUnits::new(&0u64.into(), &0u64.into())],
+	};
+
+	let gov_utxo = crate::init_governance::get_governance_utxo(genesis_utxo, client)
+		.await
+		.map_err(|e| JsError::from_str(e.to_string().as_str()))?;
```

Comment:
However, in this place missing governance utxo is rather internal error, I think.